File: 0094. Binary Tree Inorder Traversal/solution.c
#include <stdio.h>
#include <stdlib.h>
#include "../C_library/library.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
#define MAX_NODE_NUM 100
int* inorderTraversal(struct TreeNode* root, int* returnSize) {
    int *ret_array = malloc(MAX_NODE_NUM * sizeof(int));
    *returnSize = 0;
    struct TreeNode* stacks[MAX_NODE_NUM];
    int stack_next_ptr = 0;
    while(root != NULL || stack_next_ptr != 0) {
        while (root != NULL) {
            stacks[stack_next_ptr++] = root;
            root = root->left;
        }
        root = stacks[--stack_next_ptr];
        ret_array[(*returnSize)++] = root->val;
        root = root->right;
    }
    return ret_array;
}

int main() {
    int nums[] = {1,-1,2,3};
    int returnSize;
    struct TreeNode* root = genTreeFromArray(nums, sizeof(nums)/sizeof(nums[0]), -1);
    //printTreeNodes(root);
    int *ret_array = inorderTraversal(root, &returnSize);
    for(int i = 0; i < returnSize; i++)
        printf("%d ", ret_array[i]);
    printf("\n");
    freeTreeNodes(root);
    return 0;
}

File: 0094. Binary Tree Inorder Traversal/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
	fn traverse(root: Option<Rc<RefCell<TreeNode>>>, v: &mut Vec<i32>) {
	    if root == None {
	        return
	    }
	    // as_ref() and unwrap() will strip Option with reference, borrow() is for RefCell, and clone() will is to match Rc()
	    Self::traverse(root.as_ref().unwrap().borrow().left.clone(), v);
	    v.push(root.as_ref().unwrap().borrow().val);
	    Self::traverse(root.unwrap().borrow().right.clone(), v);
	}
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut v = Vec::new();
	    // To call traverse, need to add Self::
        Self::traverse(root, &mut v); // v need to define as mut ref
        v
    }
}